client/allocrunner/csi_hook_test.go

```diff
@@ -0,0 +1,253 @@
+
+func TestCSIHook(t *testing.T) {
+
+	alloc := mock.Alloc()
+	logger := testlog.HCLogger(t)
+
+	testcases := []struct {
+		name                   string
+		volumeRequests         map[string]*structs.VolumeRequest
+		expectedMounts         map[string]*csimanager.MountInfo
+		expectedMountCalls     int
+		expectedUnmountCalls   int
+		expectedClaimCalls     int
+		expectedUnpublishCalls int
+	}{
+
+		{
+			name: "simple case",
+			volumeRequests: map[string]*structs.VolumeRequest{
+				"vol0": {
+					Name:           "vol0",
+					Type:           structs.VolumeTypeCSI,
+					Source:         "testvolume0",
+					ReadOnly:       true,
+					AccessMode:     structs.CSIVolumeAccessModeSingleNodeReader,
+					AttachmentMode: structs.CSIVolumeAttachmentModeFilesystem,
+					MountOptions:   &structs.CSIMountOptions{},
+					PerAlloc:       false,
+				},
+			},
+			expectedMounts: map[string]*csimanager.MountInfo{
+				"vol0": &csimanager.MountInfo{Source: fmt.Sprintf(
+					"test-alloc-dir/%s/testvolume0/ro-file-system-single-node-reader-only", alloc.ID)},
+			},
+			expectedMountCalls:     1,
+			expectedUnmountCalls:   0, // not until this is done client-side
+			expectedClaimCalls:     1,
+			expectedUnpublishCalls: 1,
+		},
+
+		{
+			name: "per-alloc case",
+			volumeRequests: map[string]*structs.VolumeRequest{
+				"vol0": {
+					Name:           "vol0",
+					Type:           structs.VolumeTypeCSI,
+					Source:         "testvolume",
+					ReadOnly:       true,
+					AccessMode:     structs.CSIVolumeAccessModeSingleNodeReader,
+					AttachmentMode: structs.CSIVolumeAttachmentModeFilesystem,
+					MountOptions:   &structs.CSIMountOptions{},
+					PerAlloc:       true,
+				},
+			},
+			expectedMounts: map[string]*csimanager.MountInfo{
+				"vol0": &csimanager.MountInfo{Source: fmt.Sprintf(
+					"test-alloc-dir/%s/testvolume/ro-file-system-single-node-reader-only", alloc.ID)},
+			},
+			expectedMountCalls:     1,
+			expectedUnmountCalls:   0, // not until this is done client-side
+			expectedClaimCalls:     1,
+			expectedUnpublishCalls: 1,
+		},
+
+		{
+			name: "one source volume mounted two ways",
+			volumeRequests: map[string]*structs.VolumeRequest{
+				"vol0": {
+					Name:           "vol0",
+					Type:           structs.VolumeTypeCSI,
+					Source:         "testvolume0",
+					ReadOnly:       true,
+					AccessMode:     structs.CSIVolumeAccessModeSingleNodeReader,
+					AttachmentMode: structs.CSIVolumeAttachmentModeFilesystem,
+					MountOptions:   &structs.CSIMountOptions{},
+					PerAlloc:       false,
+				},
+				"vol1": {
+					Name:           "vol1",
+					Type:           structs.VolumeTypeCSI,
+					Source:         "testvolume0",
+					ReadOnly:       false,
+					AccessMode:     structs.CSIVolumeAccessModeSingleNodeWriter,
+					AttachmentMode: structs.CSIVolumeAttachmentModeFilesystem,
+					MountOptions:   &structs.CSIMountOptions{},
+					PerAlloc:       false,
+				},
+			},
```

Comment:
This is not what's being tested, but just from my own personal curiosity and understanding of CSI, is this a valid real-world scenario?

Both requests are mounting `testvolume0`, but one is a single writer and the other is a single reader, so maybe that makes it OK? 😅

Reply:
It's allowed by the client because the client doesn't enforce these. It is intentional that it's a single volume being mounted two ways, but you're right we should probably only verify compatible values here.

Reply:
Actually, I'm realizing you _can't_ have different access modes, so that's meaningless. I think you can mount the same volume multiple times with the same access mode (ex. two mounts for read-only), but that's actually a bug in the client. 😦 

I'll comment out this test for now and open an issue for it.